modules/swagger-core/src/main/java/io/swagger/jackson/ModelResolver.java

```diff
@@ -202,8 +219,11 @@ public Schema resolveProperty(JavaType propType,
 
                     if("object".equals(mi.getType())) {
                         // create a reference for the property
-                        final BeanDescription beanDesc = _mapper.getSerializationConfig().introspect(propType);
-                        String name = _typeName(propType, beanDesc);
+                        String name = mi.getName();
+                        if (StringUtils.isEmpty(name)) {
```

Comment:
I believe it should be acceptable to trust the name if its been set in the `mi`.

However, I'm not 100% confident because I'm not aware of the intention behind the below re-introspection of `propType` to get `name`. Also the tests still pass if I completely remove this `if` and just rely on `mi.getName()`.

This feature does require using the name from `mi` and not the one from `_typeName(propType, beanDesc)` as it would be wrong for classes using `@JsonValue`.